llama.cpp

```diff
@@ -1828,19 +1826,149 @@ struct llama_model_loader {
         }
     }
 
-    std::string get_arch_name() const {
-        const auto kv = LLM_KV(LLM_ARCH_UNKNOWN);
+    private:
+    template <typename T> struct gk_get_arrlen  { T & output; };
+    template <typename TI, typename TO> struct gk_set_literal { TI & input; TO & output; };
+    template <typename T>
+    void gk_set(int kid, T & result) {
```

Comment:
If you want to prevent this fallback from being used at compile time, you can just `= delete` it. Same with the other functions that throw runtime errors.

Reply:
> If you want to prevent this fallback from being used at compile time, you can just `= delete` it. 

The seems to make everything blow up. Apparently the compiler can't prove that paths like assigning an int to a `std::string &` are unreachable. That's sort of why I did the template stuff in the first place. My initial version was just comparing the typeids and I assumed that would be good enough, and unfortunately C++11 doesn't have `if constexpr`.

I think trying to switch most of the `gk_set`s to simple overloads would probably run into the same issue, since the compiler doesn't know those paths are unreachable.

Thanks for the feedback, I appreciate it. I'm going to bed soon, but when I wake up I'll try to apply more of your suggestions. Also, please let me know if there's anything I can help you with, test, whatever.

Reply:
Template are compile-time dispatch, not runtime dispatch. If using `= delete` gives you errors then there are bugs in your code that cause the wrong specialization to be selected, and the errors are correct.

Reply:
> then there are bugs in your code that cause the wrong specialization to be selected, and the errors are correct.

Unfortunately, I'm still having some trouble understanding this because I don't see where the bug is. A more concrete example would help me a lot.

With `bool get_key(const std::string & key, T & result, const bool required = false)` possibly ending up at

```c++
switch (po.tag) {
    case LLAMA_KV_OVERRIDE_INT:   gk_set_lit(po.int_value,   result); break;
    case LLAMA_KV_OVERRIDE_FLOAT: gk_set_lit(po.float_value, result); break;
    case LLAMA_KV_OVERRIDE_BOOL:  gk_set_lit(po.bool_value,  result); break;
    default: GGML_ASSERT(false && "Impossible: Unhandled override tag type");
}
```

which is one case where deleting the fallback (`gk_set_lit`) leads to the type of compile error we're talking about, you're saying there's a world where 

```c++
void gk_set_lit(const TI i, TO o) {
        (void)i; (void)o;
        throw std::runtime_error(format("gk_set_lit can't handle types: in=%s, out=%s",
            typeid(TI).name(), typeid(TO).name()));
    }
```

actually runs and throws that runtime error?

In other words, what `T` can I use with `get_key` which will lead to a branch actually being taken where the runtime error is thrown? (Just want to me clear, I'm not defending the code or saying it's good or anything, just trying to understand how it can fail in terms of behavior.)

Reply:
It's not that it would actually run, but that C++ gives you the tools to guarantee at compile-time that function is never called, instead of having to assert at runtime. C++ code that checks template types at runtime is unidiomatic.

Reply:
Thanks, got it. I thought you were saying it would actually have buggy behavior at runtime.

I'm working on a (I hope) improved version.